### PR TITLE
Write documentation

### PR DESCRIPTION
**Pull request by apmasell**

### CODE REVIEW

File: admin-guide.md [hunk 1]
```diff
@@ -0,0 +1,239 @@
...
+
+## Building Workflows
+TODO -- this is not implemented yet
+
+## Running Workflows Manually
+The command line interface allows manually running workflows. See
+`/srv/vidarr/cli run --help` for details. Normally, the caller provides 3
+pieces of information:
+
+- the argument to the workflow
```
Comment (djcooke) [inline admin-guide.md:187]: arguments*?

File: admin-guide.md [hunk 2]
```diff
@@ -0,0 +1,239 @@
...
+- the external information to associated with the output of the workflow (metadata)
```
Comment (djcooke) [inline admin-guide.md:188]: associate*

File: admin-guide.md [hunk 3]
```diff
@@ -0,0 +1,239 @@
...
+- extra parameters to provide to the workflow engine
+
+These are provided as either JSON files or JSON arguments on the command line.
+The exact format of these files will vary depending on the workflow and the
+plugin configuration.
+
+After the workflow runs successfully, the provenane records generated by the
```
Comment (djcooke) [inline admin-guide.md:195]: provenance*

File: admin-guide.md [hunk 4]
```diff
@@ -0,0 +1,239 @@
...
+files and then runs a comparison script that checks that the calculate script's
+output matches the reference data.
+
+Every provisioned out file will be symlinked into a temporary directory. The
+`"metrics_calculate"` script will be run in that directory receiving the path
+to that directory as the first argument. Anything it writes to standard output
+will be captured. Then `"metrics_calculate"` will be called with the
+`"output_metrics"` as the first argument and the file generated in the previous
+step as the second argument. If it exists non-zero, then the test is allowed to
+pass.
```
Comment (djcooke) [inline admin-guide.md:239]: exits*? Script exiting with non-zero is good?

File: architecture.md [hunk 5]
```diff
@@ -0,0 +1,98 @@
...
+Each workflow goes through three _phases_:
+
+1. Input files are provisioned in to be available to the workflow.
+2. The workflow is run.
+3. The output from the workflow is copied out to a permanent store (provision
+	 out).
+
+Every workflow definition in Vidarr has a description of the _parameters_ of
+the workflow, but _files_ are handled specially. Every workflow definition also
+has _outputs_. When it comes to to run a workflow, plugins will do the work of
```
Comment (djcooke) [inline architecture.md:13]: time to*

File: architecture.md [hunk 6]
```diff
@@ -0,0 +1,98 @@
...
+provisioning in files, running the workflow, and provisioning out the results.
+A combination of plugins is called a _target_.
+
+Getting a file into a workflow is a bit of a challenge. If the workflow is
+running on share disk, then the input needed is just a file path, but it be an
+iRODS identifier or S3 URL or something else. Vidarr provides identifiers for
```
Comment (djcooke) [inline architecture.md:19]: ...just a file path, ~~but it~~ be **it** an iRODS identifier...

File: architecture.md [hunk 7]
```diff
@@ -0,0 +1,98 @@
...
+files it has produced so any file generated by one workflow can be used as
+input for another. Plugins are responsible for figuring out how to interpret
+these identifiers. Plugins can also provision in external files and they plugin
```
Comment (djcooke) [inline architecture.md:22]: **the** plugin

File: architecture.md [hunk 8]
```diff
@@ -0,0 +1,98 @@
...
+defines what information it needs. Therefore, to run a workflow, the client
+needs to know what the _parameters_ of the workflow are and what the provision
+in information for the _target_ is.
+
+Similarly, the provision out plugins need additional information about how to
+store the output file and each plugin will have different requirements.
+
+Even running a workflow can be slightly different since the workflow engine (or
+its back-end) may have adjustable settings, such a priority.
```
Comment (djcooke) [inline architecture.md:31]: such **as** priority

File: architecture.md [hunk 9]
```diff
@@ -0,0 +1,98 @@
...
+2. Download all the workflows available on a Vidarr server.
+3. Combine workflow and targets to determine exactly what information is needed.
+4. Submit a workflow run request with all the information needed.
+
+The big goal of Vidarr is to be able to associate these files with external
+metadata information (from LIMS). Each external ID comes with a two part
+identifier (a _provider_ and an _identifier_) and version information. The
+versions will be discussed later.
+
+When an external file is feed to a provision-in plugin, it must be manually
```
Comment (djcooke) [inline architecture.md:45]: fed*

File: architecture.md [hunk 10]
```diff
@@ -0,0 +1,98 @@
...
+5. Shesmu will send `labdata/sample` with `lims-version` = `L2` and
+	 `shesmu-signature` = `S1` to Vidarr.
+6. Vidarr will recognise that while it doesn't have `lims-version` = `L2`, it
+	 does have `shesmu-signature` = `S1`. It takes this as proof that `L1` and
+   `L2` are equivalent and adds `L2` to its collection.
+
+This could also work the other way around: if the olive changes and the
+signature changes, the LIMS-provided versions will allow Vidarr to recognise
+the Shesmu keys as equivalent. It also allows the LIMS provider to change its
+version computation algorithm: if a `lims-verion-2nd-edition` version key comes
```
Comment (djcooke) [inline architecture.md:89]: `lims-version-2nd-edition`*

File: plugin-guide.md [hunk 11]
```diff
@@ -0,0 +1,92 @@
+# Vidarr Plugin Developer's Guide
+Vidarr uses plugins to allow interaction with a diverse set of system. Plugins
```
Comment (djcooke) [inline plugin-guide.md:2]: systems*

File: plugin-guide.md [hunk 12]
```diff
@@ -0,0 +1,92 @@
...
+provide multiple. Each has two classes: a small "loader" class responsible for
+loading this plugin's configuration from JSON files and a "real" interface that
+does most of the work. For loading, each plugin must provide a unique name that
+will be used in the `"type"` property of JSON configuration files.
+
+These are high-level overviews of the purpose and general constraints for each
+service. The JavaDoc for each interface provides the details for how the
+interfaces should behave.
+
+Additionally, plugins communicate with the outside world though the types they
```
Comment (djcooke) [inline plugin-guide.md:21]: through*

File: plugin-guide.md [hunk 13]
```diff
@@ -0,0 +1,92 @@
...
+expect. A description of the types is provided in
+`ca.on.oicr.gsi.vidarr.SimpleType` and the format for the values is mean to be
```
Comment (djcooke) [inline plugin-guide.md:23]: meant*

File: plugin-guide.md [hunk 14]
```diff
@@ -0,0 +1,92 @@
...
+future asynchronous tasks. Plugins must implement recovery from crash, so are
+expected to journal their current state to the database. The `WorkMonitor`
+provides methods to journal state to the database for crash recovery and to
+provide status information to users.
+
+# Input Provisioners
+Input provisioners implement `ca.on.oicr.gsi.vidarr.InputProvisionerProvider`
+and `ca.on.oicr.gsi.vidarr.InputProvisioner`. These plugins are responsible for
+taking files from existing workflows or provided by the user and generating a
+file path that a workflow can be used.
```
Comment (djcooke) [inline plugin-guide.md:37]: "that a workflow can use."

File: plugin-guide.md [hunk 15]
```diff
@@ -0,0 +1,92 @@
...
+Workflow engines implement `ca.on.oicr.gsi.vidarr.WorkflowEngineProvider`
+and `ca.on.oicr.gsi.vidarr.WorkflowEngine`. These plugins are responsible for
+running workflows and collecting the output from the workflow.
+
+The class `BaseJsonWorkflowEngine` is a partial implementation that can store
+crash recovery information in a JSON object of the implementor's choosing,
+making recovery easier.
+
+# Unloader
+Unloader implement `ca.on.oicr.gsi.vidarr.UnloaderProvider`
```
Comment (djcooke) [inline plugin-guide.md:88]: Unloaders*

File: vidarr-cromwell/README.md [hunk 16]
```diff
@@ -0,0 +1,117 @@
+# Vidarr Plugins for Cromwell
+These plugins allow interfacing Vidarr with [Cromwell](https://cromwell.readthedocs.io/en/stable/).
+
+## Cromwell Output Provisioner
+The output provision uses a Cromwell job to copy an output from a file off
```
Comment (djcooke) [inline vidarr-cromwell/README.md:5]: provisioner*

File: vidarr-cromwell/README.md [hunk 17]
```diff
@@ -0,0 +1,117 @@
...
+shared disk to another directory for permanent archival. The workflow must
+compute the file size and MD5 of the files contents and report them plus the
```
Comment (djcooke) [inline vidarr-cromwell/README.md:7]: file's* contents

File: vidarr-cromwell/README.md [hunk 18]
```diff
@@ -0,0 +1,117 @@
...
+      "type": "cromwell",
+      "wdlVersion": "1.0",
+      "workflowOptions": {
+        "read_from_cache": false,
+        "write_to_cache": false
+      },
+      "workflowUrl": "http://example.com/provisionFileOut.wdl"
+    }
+
+The "cromwellUrl"` is the Cromwell server that will handle these requests. The
```
Comment (djcooke) [inline vidarr-cromwell/README.md:26]: Missing "`" before "cromwellUrl"

File: admin-guide.md [hunk 19]
```diff
@@ -0,0 +1,239 @@
...
+    KillMode=process
+    Restart=always
+    RestartSec=10
+    
+    [Install]
+    WantedBy=multi-user.target
+
+Create a new PostgreSQL database and role for Vidarr.
+
+Now prepare a JSON file as follows:
```
[the remark this answers is not in the document]
Comment (apmasell) [reply admin-guide.md]: It's the systemd one. Updated.

File: admin-guide.md [hunk 20]
```diff
@@ -0,0 +1,239 @@
...
+    User=debian
+    ExecStart=/srv/vidarr/run /srv/vidarr/config
+    KillMode=process
+    Restart=always
+    RestartSec=10
+    
+    [Install]
+    WantedBy=multi-user.target
+
+Create a new PostgreSQL database and role for Vidarr.
```
[the remark this answers is not in the document]
Comment (apmasell) [reply admin-guide.md:53]: Added.

File: admin-guide.md [hunk 21]
```diff
@@ -0,0 +1,239 @@
...
+This guide is broken down by these roles, but they may overlap to some degree.
+
+## System Operations Guide
+Running a Vidarr instance requires:
+
+- Java 14 or later
+- PostgreSQL 12 or later
+
+### Installing Vidarr on a Linux Instance
+Pull all the JARs required for Vidarr and its plugins into a directory, say `/srv/vidarr/jars`. It can be easiest to launch Vidarr with a shellscript as follows stored in `/srv/vidarr/run`:
```
[the remark this answers is not in the document]
Comment (apmasell) [reply admin-guide.md]: I've added some build instructions. I think the future, when we start publishing it to a Maven repo, we'll need to revisit these instructions.

**Review by djcooke — APPROVED** [review]

File: architecture.md [hunk 22]
```diff
@@ -0,0 +1,112 @@
...
+`sample1`.
+
+Every external key can be associated with multiple version names and multiple
+version values. This is designed to work with [Shemu's
+signatures](https://oicr-gsi.github.io/shesmu/tutorial.html#signature-and-signable-variables).
+The design in Vidarr is meant to work something like this:
+
+1. Shesmu ingests data from LIMS and `labdata/sample1` has version `L1`.
+2. Shesmu computes a signature `S1` and passes this data to Vidarr as
+	 `labdata/sample` with `lims-version` = `L1` and `shesmu-signature` = `S1`.
```
Comment (mlaszloffy) [inline architecture.md:77]: `labdata/sample` -> `labdata/sample1`

File: architecture.md [hunk 23]
```diff
@@ -0,0 +1,112 @@
...
+3. Something changes in LIMS and now  `labdata/sample1` has version `L2`.
+4. Shesmu ingests the new data. Suppose the data the olive is using is
+	 unchanged, so Shesmu computes `S1` again.
+5. Shesmu will send `labdata/sample` with `lims-version` = `L2` and
```
Comment (mlaszloffy) [inline architecture.md:81]: `labdata/sample` -> `labdata/sample1`

File: architecture.md [hunk 24]
```diff
@@ -0,0 +1,112 @@
...
+	 out).
+
+Every workflow definition in Vidarr has a description of the _parameters_ of
+the workflow, but _files_ are handled specially. Every workflow definition also
+has _outputs_. When it comes time to run a workflow, plugins will do the work of
+provisioning in files, running the workflow, and provisioning out the results.
+A combination of plugins is called a _target_.
+
+Getting a file into a workflow is a bit of a challenge. If the workflow is
+running on share disk, then the input needed is just a file path, but it could
```
Comment (mlaszloffy) [inline architecture.md:18]: `share` -> `shared`

**Review by mlaszloffy — APPROVED** [review]

**Pull request merged**